### PR TITLE
feat: Add Gupshup Whatsapp Channel

### LINKED ISSUE #1119: WhatsApp Gupshup Integration
** Is your feature request related to a problem? Please describe. **
This is not a problem. I would like to suggest the integration of Gupshup WhatsApp. The rates are much more attractive than Twilio, in addition to having interactive buttons in the message templates.

** Describe the solution you'd like **
Integrate gupshup.io

** Describe alternatives you've considered **
Gupshup API is much better to work with, in addition to what I already mentioned we can use our own numbers to approve with WhatsApp and not rent as in Twilio

** Additional context **
I hope you think about the positive points for this integration

### PR DESCRIPTION
**Pull request by thundersparkf**
# Pull Request Template

## Gupshup Channel Integration

Added support for Gupshup WhatsApp channel receive and send(only text). This is helpful for users with existing WhatsApp approved accounts as opposed to getting a new number through 3rd party provider.

Fixes #1119 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [ ] To add tests

## Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Help needed
- [ ] Unable to add attachment button for gupshup channel in chat window.
- [ ] Contacts not building for gupshup channel

### Any help would be much appreciated.

### CODE REVIEW

File: app/javascript/dashboard/components/widgets/ChannelItem.vue [hunk 1]
```diff
@@ -40,6 +40,10 @@
       v-if="channel.key === 'twilio'"
       src="~dashboard/assets/images/channels/twilio.png"
     />
+    <img
+      v-if="channel.key === 'gupshup'"
+      src="http://www.gupshup.io/developer/resources/img/ogp.jpg"
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/ChannelItem.vue:45]: Please use a local image, otherwise, it would break if GupShup changes the URL.

File: app/javascript/dashboard/i18n/locale/he/inboxMgmt.json [hunk 2]
```diff
@@ -127,6 +127,42 @@
           "ERROR_MESSAGE": "We were not able to authenticate Twilio credentials, please try again"
         }
       },
+      "GUPSHUP": {
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/he/inboxMgmt.json:130]: Please revert these changes, it would be auto-generated by the system.

File: app/javascript/dashboard/routes/dashboard/settings/inbox/FinishSetup.vue [hunk 3]
```diff
@@ -21,6 +21,14 @@
           >
           </woot-code>
         </div>
+        <div class="medium-6 small-offset-3">
+          <woot-code
+            v-if="isAGupshupInbox"
+            lang="html"
+            :script="gupshupCallbackURL"
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/inbox/FinishSetup.vue:28]: I'm not sure if I understand this correct. I could find the `gupshupCallbackURL` method or a computed property

Comment (pranavrajs) [reply app/javascript/dashboard/routes/dashboard/settings/inbox/FinishSetup.vue:28]: My bad, it is available on the configMixin.

File: app/javascript/dashboard/routes/dashboard/settings/inbox/channels/Gupshup.vue [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  },
+  methods: {
+    async createChannel() {
+      this.$v.$touch();
+      if (this.$v.$invalid) {
+        return;
+      }
+
+      try {
+        // eslint-disable-next-line no-shadow
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/settings/inbox/channels/Gupshup.vue:116]: Please don't disable the eslint check unless required. Can we try to fix the issue here instead?

**Review by pranavrajs — CHANGES_REQUESTED** [review]
@thundersparkf Thanks for PR. ❤️ 

I took a look at the frontend changes, there are some observations. Please take a look at the individual comments. Also, I have seen a lot of console.log statements. Please remove those as well. This should be have done by prettier/eslint config, will revisit those to see if there is a mistake in those.

@sojan-official Please take a look at the backend changes.

File: app/javascript/dashboard/i18n/locale/he/inboxMgmt.json (hunk 2 above)
Comment (thundersparkf) [reply app/javascript/dashboard/i18n/locale/he/inboxMgmt.json:130]: Cool so just the locale/en is enough?

File: app/javascript/dashboard/routes/dashboard/settings/inbox/channels/Gupshup.vue (hunk 4 above)
Comment (thundersparkf) [reply app/javascript/dashboard/routes/dashboard/settings/inbox/channels/Gupshup.vue:116]: Got it! I think I accidentally added it because the IDE recommendation in the beginning. There wasn't any issue there.

File: app/javascript/dashboard/i18n/locale/he/inboxMgmt.json (hunk 2 above)
Comment (pranavrajs) [reply app/javascript/dashboard/i18n/locale/he/inboxMgmt.json]: Yes, the rest would autogenerated by Crowdin.

**Comment by thundersparkf** [comment]
@sojan-official any update on this?

**Comment by sojan-official** [comment]
@thundersparkf  Thanks for the changes. The team is slightly caught up with release chores for `v1.18.0`. We are also still evaluating the addition of the Gupshup channel from a product standpoint. We will prioritize the decision for the July release [v1.19.0].

I will get back to you with an update by next week.

**Comment by sojan-official** [comment]
@thundersparkf  Thanks a lot for working on this.

I had a chance to talk with the team about this PR. Based on the discussion, we felt that we should tackle this with a slightly different approach.  Please follow the detailed talking points on this discussion: https://github.com/chatwoot/chatwoot/discussions/2759

Let's retain this PR until we have the provider APIs ready and continue working on this from then. 

In the future, please ensure the go-ahead from a maintainer before you start the work so that we could ensure quicker merges.

**Comment by thundersparkf** [comment]
@sojan-official Thank you! I went through the discussion and it makes sense. To be honest I wasn't actually sure I could pull it off so I waited until last minute but since I saw the open-for-prs label I assumed it's cool. But thanks for the advice I'll keep it in mind.

Also, just to clarify, if we have a bridge server between chatwoot and gupshup, its possible to implement gupshup through the API channel? Would that be correct?

**Comment by sojan-official** [comment]
@thundersparkf Thanks for taking it in good spirits. Apologies for the wrong label on the original issue. 

Yes, that's totally possible in the interim.  
When you create an API inbox, you can configure a webhook URL.  This URL should be the one for your bridge server / lamda function. 

- So whenever an action occurs, chatwoot will keep sending events to your bridge. From there you can filter the relevant events like agent replies, User the Gupshup API endpoint to create the messages back to your users. 

- Now for the case where the user replies to your messages,  implement a webhook on your bridge server. parse the payload that Gupshup sends, create a message back in chatwoot using chatwoot APIs. 

- there is also an option to work with websocket API to consume events instead of the webhooks. 

do let us know if you have any questions.

**Comment by thundersparkf** [comment]
@sojan-official thanks for the idea these are helpful. Looking forward to the Provider APIs!

**Comment by sojan-official** [comment]
Closing this PR since there wasn't much activity over this for a while. Let's reopen this once this becomes a priority again.
